src/platform/cc13xx_26xx/FactoryDataProvider.cpp

```diff
@@ -0,0 +1,333 @@
+    ReturnErrorCodeIf(!mFactoryData.spake2p_verifier.data, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND);
+    memcpy(verifierBuf.data(), mFactoryData.spake2p_verifier.data, mFactoryData.spake2p_verifier.len);
+    verifierLen = mFactoryData.spake2p_verifier.len;
+    verifierBuf.reduce_size(verifierLen);
+    return CHIP_NO_ERROR;
+}
+CHIP_ERROR FactoryDataProvider::GetSetupPasscode(uint32_t & setupPasscode)
+{
+    ReturnErrorCodeIf(!mFactoryData.passcode.data, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND);
+    memcpy(&setupPasscode, mFactoryData.passcode.data, mFactoryData.passcode.len);
+    return CHIP_NO_ERROR;
+}
+CHIP_ERROR FactoryDataProvider::SetSetupPasscode(uint32_t setupPasscode)
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+CHIP_ERROR FactoryDataProvider::GetVendorName(char * buf, size_t bufSize)
+{
+    ReturnErrorCodeIf(bufSize < mFactoryData.vendor_name.len, CHIP_ERROR_BUFFER_TOO_SMALL);
+    ReturnErrorCodeIf(!mFactoryData.vendor_name.data, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND);
+    memcpy(buf, mFactoryData.vendor_name.data, mFactoryData.vendor_name.len);
+    ChipLogProgress(DeviceLayer, "vendor name: %s", buf);
+    return CHIP_NO_ERROR;
+}
+CHIP_ERROR FactoryDataProvider::GetVendorId(uint16_t & vendorId)
+{
+    ReturnErrorCodeIf(!mFactoryData.vendor_id.data, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND);
+    memcpy(&vendorId, mFactoryData.vendor_id.data, mFactoryData.vendor_id.len);
+    return CHIP_NO_ERROR;
+}
+CHIP_ERROR FactoryDataProvider::GetProductName(char * buf, size_t bufSize)
+{
+    ReturnErrorCodeIf(bufSize < mFactoryData.product_name.len, CHIP_ERROR_BUFFER_TOO_SMALL);
+    ReturnErrorCodeIf(!mFactoryData.product_name.data, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND);
+    memset(buf, 0, bufSize);
+    memcpy(buf, mFactoryData.product_name.data, mFactoryData.product_name.len);
+    return CHIP_NO_ERROR;
+}
+CHIP_ERROR FactoryDataProvider::GetProductId(uint16_t & productId)
+{
+    ReturnErrorCodeIf(!mFactoryData.product_id.data, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND);
+    memcpy(&productId, mFactoryData.product_id.data, mFactoryData.product_id.len);
+    return CHIP_NO_ERROR;
+}
+CHIP_ERROR FactoryDataProvider::GetPartNumber(char * buf, size_t bufSize)
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+CHIP_ERROR FactoryDataProvider::GetProductURL(char * buf, size_t bufSize)
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+CHIP_ERROR FactoryDataProvider::GetProductLabel(char * buf, size_t bufSize)
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR FactoryDataProvider::GetSerialNumber(char * buf, size_t bufSize)
+{
+    ReturnErrorCodeIf(bufSize < mFactoryData.serial_number.len, CHIP_ERROR_BUFFER_TOO_SMALL);
+    ReturnErrorCodeIf(!mFactoryData.serial_number.data, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND);
+    memset(buf, 0, bufSize);
+    memcpy(buf, mFactoryData.serial_number.data, mFactoryData.serial_number.len);
```

Comment:
Does `mFactoryData.serial_number.len` include the null terminator?

Reply:
This length does not include the null terminator. The factory data creation process takes into account how the serial number needs to be formatted such that it is copied in properly to the buffer. More documentation on this process can be found in docs/guides/ti

Reply:
Caller expects a null-terminated output buffer, but this function is not guaranteeing that, as far as I can see.